pcdet/models/dense_heads/anchor_head_template.py

```diff
@@ -101,38 +101,53 @@ def assign_targets(self, gt_boxes):
+            box_cls_labels, n_labels=self.num_class[0], dtype=cls_preds.dtype
+        )
+        cls_weights = self._get_cls_weights(box_cls_labels)
         batch_size = int(cls_preds.shape[0])
-        cared = box_cls_labels >= 0  # [N, num_anchors]
+        cls_preds = cls_preds.view(batch_size, -1, self.num_class[0])
+        cls_loss = self._get_weighted_loss(
+            cls_preds, one_hot_targets, cls_weights, batch_size
+        )
+        tb_dict = {
+            'rpn_loss_cls': cls_loss.item()
+        }
+        return cls_loss, tb_dict
+
+    def _get_cls_weights(self, box_cls_labels):
         positives = box_cls_labels > 0
         negatives = box_cls_labels == 0
         negative_cls_weights = negatives * 1.0
         cls_weights = (negative_cls_weights + 1.0 * positives).float()
-        reg_weights = positives.float()
-        if self.num_class == 1:
-            # class agnostic
-            box_cls_labels[positives] = 1
+        positives_normalizer = positives.sum(1, keepdim=True).float()
+        cls_weights /= torch.clamp(positives_normalizer, min=1.0)
+        return cls_weights
+
+    def _get_weighted_loss(self, cls_preds, one_hot_targets, cls_weights, batch_size):
+        cls_loss_src = self.cls_loss_func(
+            cls_preds, one_hot_targets, weights=cls_weights
+        )  # [N, M]
+        cls_loss = cls_loss_src.sum() / batch_size
+        cls_loss = cls_loss * self.model_cfg.LOSS_CONFIG.LOSS_WEIGHTS["cls_weight"]
+        return cls_loss
 
-        pos_normalizer = positives.sum(1, keepdim=True).float()
-        reg_weights /= torch.clamp(pos_normalizer, min=1.0)
-        cls_weights /= torch.clamp(pos_normalizer, min=1.0)
+    def _get_one_hot_box_labels(self, box_cls_labels, n_labels, dtype=torch.float32):
+        if n_labels == 1:
+            # class agnostic
+            box_cls_labels[box_cls_labels > 0] = 1
+        cared = box_cls_labels >= 0  # [N, num_anchors]
         cls_targets = box_cls_labels * cared.type_as(box_cls_labels)
         cls_targets = cls_targets.unsqueeze(dim=-1)
 
         cls_targets = cls_targets.squeeze(dim=-1)
         one_hot_targets = torch.zeros(
-            *list(cls_targets.shape), self.num_class + 1, dtype=cls_preds.dtype, device=cls_targets.device
+            *list(cls_targets.shape), n_labels + 1, dtype=dtype, device=cls_targets.device
         )
         one_hot_targets.scatter_(-1, cls_targets.unsqueeze(dim=-1).long(), 1.0)
-        cls_preds = cls_preds.view(batch_size, -1, self.num_class)
         one_hot_targets = one_hot_targets[..., 1:]
-        cls_loss_src = self.cls_loss_func(cls_preds, one_hot_targets, weights=cls_weights)  # [N, M]
-        cls_loss = cls_loss_src.sum() / batch_size
 
-        cls_loss = cls_loss * self.model_cfg.LOSS_CONFIG.LOSS_WEIGHTS['cls_weight']
-        tb_dict = {
-            'rpn_loss_cls': cls_loss.item()
-        }
-        return cls_loss, tb_dict
+        return one_hot_targets
```

Comment:
In this file you did a quite large refactor. It's quite difficult to see what is different with respect to the previous version in the diff and to spot errors. In general I'm against large modifications to a fork that we barely maintain because I would like to minimise the number of differences with respect to the main trunk, in particular in files containing important logic.

Can you rewrite the code minimising the number of lines changed?

Reply:
I've already done that, and this is the best I can do. It moves some logic from the loss function into separate functions so they can be re-used in `AnchorHeadWithType`.

Reply:
Actually you're right, I took over code from Nora but she divided more functions than really necessary. Fixed it now.

Reply:
Much better now!